libraries/CurieBLE/src/BLECentralRole.cpp

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright (c) 2016 Intel Corporation.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ */
+
+#include "BLECentralRole.h"
+
+
+void ble_central_device_found(const bt_addr_le_t *addr, 
+                              int8_t rssi, 
+                              uint8_t type,
+                              const uint8_t *ad, 
+                              uint8_t len)
+{
+	char dev[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(addr, dev, sizeof(dev));
+	pr_debug(LOG_MODULE_BLE, "[DEVICE]: %s, AD evt type %u, AD data len %u, RSSI %i\n",
+	       dev, type, len, rssi);
+
+    BLECentralRole::instance()->handleDeviceFound(addr, rssi, type,
+                                                  ad, len);
+}
+
+
+BLECentralRole* BLECentralRole::_ble_central_ins = NULL;
+
+BLECentralRole *BLECentralRole::instance()
+{
+    if (NULL == _ble_central_ins)
+    {
+        _ble_central_ins = new BLECentralRole();
```

Comment:
Yes, it call the malloc and get the memory resource. But the stack and Nordic’s code didn’t have API to deregister the BLE profile. And the stack also use the pointer that point to the CurieBLE LIB. If we release those resource, the stack will have wild pointer and we didn't have a API to clean it. So we can't release those resource. Our current design use singleton pattern. Those resource can be managed and never leak.